dev/doc/changes.md

```diff
@@ -43,6 +43,19 @@ Coqlib:
   command then enables to locate the registered constant through its name. The
   name resolution is dynamic.
 
+Proof state:
+
+- Handling of proof state has been fully functionalized, thus it is
+  not possible to call global functions such as `get_current_context ()`.
+
+  The main type for functions that need to handle proof state is
+  `Lemmas.t`, for proofs that intend to associate a constant.
```

Comment:
I've nothing against having the proof state being define in the module Lemmas, but having it names `Lemmas.t` is not a good idea. Either put it in a sub module `ProofState` or call it `proof_state`.

Reply:
To clarify, `Lemmas.t` is also for "ongoing proofs" not just for finished ones.

Reply:
Note that we are not just moving proof state here, but the complete information relevant for "Lemma-like" constants which can have obligations attached, etc...

If you keep that in mind I think that `Lemmas.t` is not so bad, indeed it is an object that contains a Lemma! I am open to other namings of course but I am not sure, with that in mind, that "proof state" is accurate or btter.

Reply:
So we will have more or less:
- `Proof.t`: "proof objects", that as of today can have several entries but that's a bit of a mess.
- `Declaration.t`: Declaration-like objects, this can contain several constants, and hooks, they extend `Proof.t`
- `Lemma.t`: extends `Declaration.t` and contains a few more lemma-specific information; maybe indeed it should be merged with `Declaration.t`, however the codepaths are duplicated so even if the current series of PRs improve the situation it will take more effort.

Reply:
I am defending thus the terminology where we reserve "proof" for proof objects attached or not to constants.

Reply:
@gares I've fixed the changes entry tho, which indeed was confusing and incorrect.

Reply:
> To clarify, `Lemmas.t` is also for "ongoing proofs" not just for finished ones.

Good point, `Lemmas.t` could contain a finished proof associated to a constant, but somehow not yet propagated to a certain part of the env.

Reply:
In `Lemmas.mli` there are API getting arguments of type `Lemma.t` with the label `~pstate`, eg the ones for saving a proof. I guess they should take a `~lemma` now?

Anyway, the doc is more clear now and it explains why `Lemma.t` is not just a renaming of `Proof_global.t`. Still it represent an ongoing proof (of a lemma). So to me the name could still be proof state, even if it is larger than the old proof state. Let's see what other think about the new names.

Reply:
Good point about the `pstate` name, I did not rename as to easy porting, but I agree that I should rename at least a few. In some cases it should be `~lemma`, in some others I guess `~ontop` is still the right choice.

Reply:
I like the `Lemma.t` as for example we can provide `Lemma.equal` etc..., and it is shorter, but `Lemma.proof` etc... could also work; I dunno, indeed let's hear feedback.

Reply:
The thing is that I understand `Lemma.proof` as a projection `proof : Lemma.t -> Proof.t`

Reply:
Also `start_proof` and friends should likely be named `start_lemma`; I didn't rename as to ease diffs, but I am OK with renaming if we think it is worth it.

Reply:
@gares I've done a pass on argument naming, I've also removed some spurious uses of `Lemma.t` where it should be `Proof_global.t`; before things where hard to see with the huge dependency.